### PR TITLE
Article: Showboaters, Maxmialists and You

### PR DESCRIPTION
**Pull request by adamgordonbell**
Preview URL:
https://65b9140a3edd880094a9234f--ephemeral-narwhal-3ac3b2.netlify.app/blog/showboaters/

### CODE REVIEW

File: blog/_posts/2024-01-30-showboaters.md [hunk 1]
```diff
@@ -0,0 +1,269 @@
...
+This is so much nicer—less boilerplate to write and fewer places for bugs to hide. I feel the same way about [if expressions](/single-expression-functions).
+
+~~~{.scala caption="If with assignment in branches"}
+if( x > 7){
+  y = 5
+} elseif (x > 5) {
+  y = 4
+} else {
+  y' = 3
+}
```
Comment (vladaionescu) [inline blog/_posts/2024-01-30-showboaters.md:69]: The various whitespace inconsistencies are distracting here. Might be worth fixing :-)

File: blog/_posts/2024-01-30-showboaters.md [hunk 2]
```diff
@@ -0,0 +1,269 @@
...
+
+This is so much nicer—less boilerplate to write and fewer places for bugs to hide. I feel the same way about [if expressions](/single-expression-functions).
+
+~~~{.scala caption="If with assignment in branches"}
+if( x > 7){
+  y = 5
+} elseif (x > 5) {
+  y = 4
+} else {
+  y' = 3
```
Comment (vladaionescu) [inline blog/_posts/2024-01-30-showboaters.md:68]:
```suggestion
  y = 3
```

File: blog/_posts/2024-01-30-showboaters.md [hunk 3]
```diff
@@ -0,0 +1,269 @@
...
+~~~
+
+~~~{.scala caption="If expression"}
+y = if( x > 7){
+  5
+} elseif (x > 5) {
+  4 
+} else {
+  3
+}
```
Comment (vladaionescu) [inline blog/_posts/2024-01-30-showboaters.md:79]: Similarly here

File: blog/_posts/2024-01-30-showboaters.md [hunk 4]
```diff
@@ -0,0 +1,269 @@
...
+~~~
+
+And my greeter could be something like:
+
+~~~{.java caption="More Java with Lambda"}
+
+Greeter greeter = () -> System.out.println("Hello, world!");
+~~~
+
+This is so much nicer—less boilerplate to write and fewer places for bugs to hide. I feel the same way about [if expressions](/single-expression-functions).
```
Comment (vladaionescu) [inline blog/_posts/2024-01-30-showboaters.md:60]: The link is a 404 for me.

File: blog/_posts/2024-01-30-showboaters.md [hunk 5]
```diff
@@ -0,0 +1,269 @@
...
+} elseif (x > 5) {
+  4 
+} else {
+  3
+}
+~~~
+
+There are some redundancies in the first, and because of that, it's possible to have errors. You learn to skim over the boilerplate parts of code, but those parts can have bugs, hence the error in the first version. The slightly shorter expression-based code has less room for error.
+
+I'm not a C++ programmer, but C++'s `constexpr` is another great example of improved expressiveness and readability. Doing things at compile time previously involved templates, but now you can use `constexpr`:
```
Comment (vladaionescu) [inline blog/_posts/2024-01-30-showboaters.md:84]:
```suggestion
I'm not a C++ programmer, but C++'s `constexpr` is another great example of improved expressiveness and readability. Doing things at compile time previously involved macros, but now you can use `constexpr`:
```

File: blog/_posts/2024-01-30-showboaters.md [hunk 6]
```diff
@@ -0,0 +1,269 @@
...
+    .liftM[RequestsState]
+  _ <- askFetchJudge.forever
+} yield ()
+~~~
+
+<figcaption>Call a Weather API( from [Pawel Szulc excellent talk](https://www.youtube.com/watch?v=y_QHSDOVJM8) ) </figcaption>
+
+This example is a little over the top, but the reaction is a real thing that happens when someone first encounters code from a language using complex syntax, advanced concepts, and concise code all at once.
+
+This isn't just an FP thing, either. Here's some Go code for [leet code 2529](https://leetcode.com/problems/maximum-count-of-positive-integer-and-negative-integer/description/) :
```
Comment (vladaionescu) [inline blog/_posts/2024-01-30-showboaters.md:152]: Spell out "Functional Programming" for clarity?

File: blog/_posts/2024-01-30-showboaters.md [hunk 7]
```diff
@@ -0,0 +1,269 @@
...
+
+~~~{.cpp caption="Code Report: Count Max Rust"}
+pub fn maximum_count(nums: Vec<i32>) -> i32 {
+    let pos = nums.clone().into_iter().filter(|e| *e > 0).count() as i32;
+    let neg = nums.into_iter().filter(|e| *e < 0).count() as i32;
+    std::cmp::max(pos, neg)
+}
+~~~
+
+This Scala, C++, and Rust code may all cause a particular reaction in the unfamiliar: 'why all the concepts used, when the solution could be so simple'. But the examples are quite different. The Scala one is showboating, and others are not.
```
Comment (vladaionescu) [inline blog/_posts/2024-01-30-showboaters.md:197]: LoL

File: blog/_posts/2024-01-30-showboaters.md [hunk 8]
```diff
@@ -0,0 +1,269 @@
...
+
+And the Rust solution requires knowledge of iterators, filters, and more.
+
+~~~{.cpp caption="Code Report: Count Max Rust"}
+pub fn maximum_count(nums: Vec<i32>) -> i32 {
+    let pos = nums.clone().into_iter().filter(|e| *e > 0).count() as i32;
+    let neg = nums.into_iter().filter(|e| *e < 0).count() as i32;
+    std::cmp::max(pos, neg)
+}
+~~~
```
Comment (vladaionescu) [inline blog/_posts/2024-01-30-showboaters.md:195]: Side note: if you talk to people who train for programming competitions, they will have the weirdest ways to express simple things like this. They optimize for what is easiest to type - they'll use super short variable names, and contrived techniques that shorten the number of keystrokes necessary to write out the code live to the minimum, while not sacrificing performance. It looks super weird - border-line obfuscated code.

**Review by vladaionescu — APPROVED** [review]
Very nice hehe. Just had some minor remarks.

File: blog/_posts/2024-01-30-showboaters.md (hunk 2 above)
Comment (adamgordonbell) [reply blog/_posts/2024-01-30-showboaters.md:68]: I put this typo in on purpose, to show how the verbosity could hide problems.

**Pull request merged**